common/flatpak-dir.c

```diff
@@ -7793,13 +7965,83 @@ flatpak_dir_install (FlatpakDir          *self,
         }
       else
         {
-          /* We're pulling from a remote source, we do the network mirroring pull as a
-             user and hand back the resulting data to the system-helper, that trusts us
-             due to the GPG signatures in the repo */
+          g_autoptr (GError) local_error = NULL;
+          g_autofree gchar *src_dir = NULL;
+          g_autofree gchar *mountpoint = NULL;
+          g_autoptr(GSubprocess) revokefs_fuse = NULL;
+          gint socket = -1;
+
+          if (!flatpak_dir_system_helper_call_get_revokefs_fd (self,
+                                                               FLATPAK_HELPER_GET_REVOKEFS_FD_FLAGS_NONE,
+                                                               ref,
+                                                               opt_commit,
+                                                               installation ? installation : "",
+                                                               &socket,
+                                                               &src_dir,
+                                                               cancellable,
+                                                               &local_error))
+            {
+              if (g_error_matches (local_error, G_DBUS_ERROR, G_DBUS_ERROR_NOT_SUPPORTED))
+                g_debug ("revokefs-fuse not supported on your installation: %s", local_error->message);
+              else
+                g_warning ("Failed to get revokefs socket from system-helper: %s", local_error->message);
+            }
+          else
+            {
+              g_autoptr(GSubprocessLauncher) launcher = NULL;
+              g_autoptr(GFile) mountpoint_file = NULL;
+              g_autofree gchar *client_uid = NULL;
+              g_autofree gchar *repo_basename = NULL;
 
-          child_repo = flatpak_dir_create_system_child_repo (self, &child_repo_lock, NULL, error);
+              mountpoint = flatpak_dir_revokefs_fuse_create_mountpoint (ref, &local_error);
+              if (mountpoint == NULL)
+                {
+                  g_warning ("Failed to create a mountpoint for revokefs-fuse: %s", local_error->message);
+                  flatpak_dir_cancel_pull (self, FLATPAK_HELPER_CANCEL_PULL_FLAGS_NONE, cancellable, src_dir);
+                  goto pull_fallback;
```

Comment:
I'm not a big fan of `goto out` (used heavily in this file) but I really don't like `goto`s which jump anywhere other than the error-recovery path at the end of a function. I think this block which could possibly fail should be factored out into a function.

Reply:
I continue to feel that this use of `goto` is not great.